File: app/src/main/java/com/pokewidgets/app/sprite/TrainerArt.kt
package com.pokewidgets.app.sprite

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.pokewidgets.app.catalog.TrainerBack

/**
 * Turns downloaded trainer images into drawable sprites.
 *
 * Showdown's fronts are ready to use. Backs come straight from game graphics, which need
 * two things done to them:
 *
 * - **Emerald and FireRed** store the throwing animation as a strip of frames; only the
 *   first, standing pose is kept.
 * - **Game Boy and Game Boy Color** sprites are stored as four shades of grey, with the
 *   colours living elsewhere in the ROM, and their background is plain white. The shades are
 *   mapped onto the game's palette and the white surrounding the figure is made transparent
 *   — only white reachable from the edge, so the whites of the eyes stay white.
 */
object TrainerArt {

    fun decode(bytes: ByteArray, back: TrainerBack?): Bitmap? {
        val options = BitmapFactory.Options().apply {
            inPreferredConfig = Bitmap.Config.ARGB_8888
            inScaled = false
        }
        val decoded = BitmapFactory.decodeByteArray(bytes, 0, bytes.size, options) ?: return null
        if (back == null) return decoded

        val w = minOf(back.frameWidth, decoded.width)
        val h = minOf(back.frameHeight, decoded.height)
        val pixels = IntArray(w * h)
        decoded.getPixels(pixels, 0, w, 0, 0, w, h)
        decoded.recycle()

        back.palette?.let { hex ->
            val palette = IntArray(hex.size) { parseColor(hex[it]) }
            recolorShades(pixels, palette)
            clearEdgeBackground(pixels, w, h, palette.first())
        }
        return Bitmap.createBitmap(pixels, w, h, Bitmap.Config.ARGB_8888)
    }

    /**
     * Maps each grey pixel to the palette entry for its shade: white, light, dark, black.
     * Pure, so the mapping is tested without Android.
     */
    fun recolorShades(pixels: IntArray, palette: IntArray) {
        if (palette.isEmpty()) return
        val last = palette.size - 1
        for (i in pixels.indices) {
            val p = pixels[i]
            if (p ushr 24 == 0) continue
            val grey = (p shr 16 and 0xFF) // greyscale: every channel is the same
            val shade = ((255 - grey) * last + 127) / 255
            pixels[i] = palette[shade.coerceIn(0, last)]
        }
    }

    /**
     * Makes [background]-coloured pixels transparent where they touch the edge of the image,
     * following them inward, and leaves enclosed ones alone.
     */
    fun clearEdgeBackground(pixels: IntArray, w: Int, h: Int, background: Int) {
        if (w <= 0 || h <= 0) return
        val stack = IntArray(w * h)
        var top = 0
        fun push(x: Int, y: Int) {
            val i = y * w + x
            if (pixels[i] == background) {
                pixels[i] = 0
                stack[top++] = i
            }
        }
        for (x in 0 until w) {
            push(x, 0)
            push(x, h - 1)
        }
        for (y in 0 until h) {
            push(0, y)
            push(w - 1, y)
        }
        while (top > 0) {
            val i = stack[--top]
            val x = i % w
            val y = i / w
            if (x > 0) push(x - 1, y)
            if (x < w - 1) push(x + 1, y)
            if (y > 0) push(x, y - 1)
            if (y < h - 1) push(x, y + 1)
        }
    }

    private fun parseColor(hex: String): Int =
        (0xFF000000.toInt()) or (hex.removePrefix("#").toLong(16).toInt() and 0xFFFFFF)
}

File: web/src/core/trainerArt.ts
import type { Trainer, TrainerBack, TrainerIndex } from './catalog';
import { trainerFrontUrl } from './catalog';
import { loadFirst, withProxy } from './imageSource';

/**
 * Turns a trainer's downloaded image into something drawable, the same way
 * `sprite/TrainerArt.kt` does:
 *
 * - Showdown's fronts are used as they are.
 * - Backs come from game graphics: Emerald and FireRed store the throw animation as a strip
 *   of frames, so only the first standing pose is kept, and Game Boy sprites are four shades
 *   of grey whose colours live elsewhere in the ROM, so they are recoloured and their white
 *   background cleared.
 *
 * Fronts come from play.pokemonshowdown.com, which sends no CORS header, so their pixels
 * cannot be read — they need no processing, and are handed back as a plain URL. The backs
 * are on jsDelivr, which does, so those can go through a canvas. Either can come through the
 * image proxy when its own host is out of reach (see `imageSource.ts`); the proxy allows
 * cross-origin reads too, so a proxied back is still recoloured.
 */

export interface TrainerArt {
  /** A URL or data URL to draw. */
  src: string;
  width: number;
  height: number;
  /** True when this is the front standing in for a back sprite no game ever drew. */
  standIn: boolean;
}

const cache = new Map<string, TrainerArt>();

const parseColor = (hex: string): [number, number, number] => {
  const value = Number.parseInt(hex.replace('#', ''), 16);
  return [(value >> 16) & 0xff, (value >> 8) & 0xff, value & 0xff];
};

/** Maps each grey to its palette entry: white, light, dark, black. */
function recolorShades(data: Uint8ClampedArray, palette: string[]): void {
  const colors = palette.map(parseColor);
  const last = colors.length - 1;
  for (let i = 0; i < data.length; i += 4) {
    if (data[i + 3] === 0) continue;
    const grey = data[i]!;
    const shade = Math.min(Math.max(Math.round(((255 - grey) * last) / 255), 0), last);
    const [r, g, b] = colors[shade]!;
    data[i] = r;
    data[i + 1] = g;
    data[i + 2] = b;
  }
}

/** Clears the background colour where it touches the edge, so the whites of eyes survive. */
function clearEdgeBackground(data: Uint8ClampedArray, w: number, h: number, background: string): void {
  const [br, bg, bb] = parseColor(background);
  const matches = (i: number) => data[i] === br && data[i + 1] === bg && data[i + 2] === bb && data[i + 3] !== 0;
  const stack: number[] = [];
  const push = (x: number, y: number) => {
    const i = (y * w + x) * 4;
    if (matches(i)) {
      data[i + 3] = 0;
      stack.push(x, y);
    }
  };
  for (let x = 0; x < w; x++) {
    push(x, 0);
    push(x, h - 1);
  }
  for (let y = 0; y < h; y++) {
    push(0, y);
    push(w - 1, y);
  }
  while (stack.length) {
    const y = stack.pop()!;
    const x = stack.pop()!;
    if (x > 0) push(x - 1, y);
    if (x < w - 1) push(x + 1, y);
    if (y > 0) push(x, y - 1);
    if (y < h - 1) push(x, y + 1);
  }
}

async function processBack(back: TrainerBack): Promise<TrainerArt> {
  const { image } = await loadFirst(withProxy([back.u]), true);
  const w = Math.min(back.w, image.naturalWidth);
  const h = Math.min(back.h, image.naturalHeight);
  const canvas = document.createElement('canvas');
  canvas.width = w;
  canvas.height = h;
  const ctx = canvas.getContext('2d')!;
  ctx.imageSmoothingEnabled = false;
  // Only the first frame of the throw animation: the trainer standing.
  ctx.drawImage(image, 0, 0, w, h, 0, 0, w, h);

  if (back.p?.length) {
    const pixels = ctx.getImageData(0, 0, w, h);
    recolorShades(pixels.data, back.p);
    clearEdgeBackground(pixels.data, w, h, back.p[0]!);
    ctx.putImageData(pixels, 0, 0);
  }
  return { src: canvas.toDataURL('image/png'), width: w, height: h, standIn: false };
}

/**
 * The trainer as they should be drawn. A back that no game ever drew comes back as the
 * front with `standIn` set, so the caller can turn it round and say so.
 */
export async function trainerArt(index: TrainerIndex, trainer: Trainer, back: boolean): Promise<TrainerArt> {
  const key = `${trainer.i}/${back}`;
  const cached = cache.get(key);
  if (cached) return cached;

  let art: TrainerArt;
  if (back && trainer.b) {
    try {
      art = await processBack(trainer.b);
    } catch {
      art = await front(index, trainer, true);
    }
  } else {
    art = await front(index, trainer, back);
  }
  cache.set(key, art);
  return art;
}

async function front(index: TrainerIndex, trainer: Trainer, standIn: boolean): Promise<TrainerArt> {
  const { src, image } = await loadFirst(withProxy([trainerFrontUrl(index, trainer)]));
  return { src, width: image.naturalWidth, height: image.naturalHeight, standIn };
}
